freud/environment.pyx

```diff
@@ -532,11 +532,56 @@ cdef class EnvironmentCluster(_MatchEnv):
                 global_search=False):
         r"""Determine clusters of particles with matching environments.
 
-        In general, it is recommended to specify a number of neighbors rather
-        than just a distance cutoff as part of your neighbor querying when
-        performing this computation. Using a distance cutoff alone could easily
-        lead to situations where a point doesn't match a cluster because a
-        required neighbor is just outside the cutoff.
+        An environment is defined by the bond vectors between a particle and its
+        neighbors as defined by :code:`env_neighbors`.
+        For example, :code:`env_neighbors= {'num_neighbors': 8}` means that every
+        particle's local environment is defined by its 8 nearest neighbors.
+        Then, each particle's environment is compared to the environments of a
+        selected list of other particles in the system, controlled by the
+        :code:`neighbors` parameter. For example, :code:`neighbors={'r_max': 3.0}`
+        means that the environment of each particle will be compared to the
+        environment of every particle within a distance of 3.0.
+
+        Two environments are compared using `point-set registration
+        <https://en.wikipedia.org/wiki/Point-set_registration)>`_.
+        Specifically, we use the `Kabsch algorithm
+        <https://en.wikipedia.org/wiki/Kabsch_algorithm>`_ to find the optimal
+        rotation matrix mapping one environment onto another. The Kabsch
+        algorithm assumes a one-to-one correspondence between points in the two
+        environments. However, we typically do not know which point in one environment
+        corresponds to a particular point in the other environment. The brute force
+        solution is to try all possible correspondences, but the resulting
+        combinatorial explosion makes this problem infeasible, so we use a tolerance
+        :code:`threshold` to terminate the search when we have found a permutation
+        of points that results in a sufficiently low RMSD.
```

Comment:
I think this is a wonderful explanation, thank you! I would say as a final sentence here: 

The thresholding criterion we apply in order to determine if the two point sets have sufficiently low RMSD (i.e. "match") is quite conservative: the two point sets match if and only if, for every pair of matched points between the sets, the distance between the matched pair is less than :code:`threshold`.

That way, there is no ambiguity about how matching works here. Otherwise the user might think that we match according to RMSD lower than the threshold, which is not actually precisely what is happening.

Reply:
Ah that is a good point. We perform brute force registration between pairs of environments, but the threshold is applied to the vector distance between each pair of vectors in the environments, not to the RMSD.

Reply:
I was looking at the code, and it seems like if `registration=False`, only the thresholding criterion @erteich mentioned above is used? i.e., we brute force all pairs of vectors of two environments and compute the squared distance. No RMSD is computed for two sets of environments.
The Kabsch algorithm is only used if `registration=True`, so RMSD is only computed and minimized to find the rotation if `registration=True`?

Reply:
You're right, the RMSD computation only happens if `registration=True`. I forgot that code path even existed. @erteich what's the value in _not_ performing registration? That code path leads to somewhat arbitrary results since it depends on the order in which the elements of the environment are traversed, which isn't great. Is it just useful for quick testing because it's faster to run without registration.

You mentioned the possibility of different similarity functions (plugging in a different `isSimilar` implementation). With vs without registration actually seems like that's exactly what it does. Maybe we should refactor the code to treat those as two different similarity functions, and in 3.0 allow users to specify which similarity function to use instead? That feels like a more flexible API.

Reply:
Hi @vyasr , hopefully the comment I just made (I actually made it last night but forgot to "submit" it, so I think it is only visible to you now?) somewhat addresses your question. You're right that it depends on the order in which the elements of the environment are traversed- it is certainly not the most ideal way to do this. Like the method using registration though, the positive side is that the thresholding is such that the match found (local minimum) satisfies a certain similarity criterion, regardless of whether it is a true global minimum or not. The downside is that there may be cases in which local minima that meet the similarity criterion exist but are not found, although my intuition is that that would be rare for most use cases. But certainly there must be much better ways of doing this- I would be interested in coding one up if it is quick-ish to do.

Regarding different similarity functions, I'm into the idea of refactoring the code! That seems cool to me, since the two methods are using fairly different algorithms, as you said. There are definitely use cases in which users would not want to use registration- for example, if users wanted to distinguish between two different crystal grains, they would be interested in environments that are the same without rotation. One set of similar environments would be one crystal grain (oriented in one direction) and the other set of similar environments would be another crystal grain (oriented in another direction).

Reply:
> That code path leads to somewhat arbitrary results since it depends on the order in which the elements of the environment are traversed, which isn't great. 

Since the "thresholding criterion" compares every single pair of vectors of 2 environments and all of them need to satisfy the criterion for the 2 environments to be considered "match", I feel like the order doesn't matter? And it seems like if `registration=False` with the same `threshold`, we can only get less matches compared to `registration=True`?

> You mentioned the possibility of different similarity functions (plugging in a different `isSimilar` implementation). With vs without registration actually seems like that's exactly what it does. Maybe we should refactor the code to treat those as two different similarity functions, and in 3.0 allow users to specify which similarity function to use instead? That feels like a more flexible API.

This probably needs further discussion, and is out of the scope of this PR. I can create another issue if that's better.

Reply:
Agreed, definitely out of scope for this PR. I'm just using these discussions as a way to chart a path forward, but we definitely don't need to address all the discussions here.

Reply:
>  You're right that it depends on the order in which the elements of the environment are traversed- it is certainly not the most ideal way to do this. 

I'm still confused about @vyasr's question on how the order of traversing the vectors affects the comparison. Since we compare every single pair of vectors of 2 environments, all possible correspondence is checked right? And all of them needs to satisfy the threshold. It's just if `registration = True`, the query environment or the motif will be rotated before comparison. So if we use the same threshold for `registration = True` and `registration = False`, 2 environments are less likely to be considered match in the latter case?

Reply:
> All possible correspondences are _checked_, but look at the code [here](https://github.com/glotzerlab/freud/blob/master/cpp/environment/MatchEnv.cc#L329). It simply checks whether two vectors are close enough, then tries to insert. That means that for a given vector `i` in the first environment, the first vector `j` that is close enough will be set as a match, even if there is another _closer_ vectors `k` that happens to be checked after it. We could modify that logic if we wanted to, although that's out of scope for this PR. I just wanted to point out the issue.

I see what you mean now. Since the problem is that once `threshold` is satisfied by the pair of vectors, the `vec_map` is no longer updated for this pair of vectors. Would this be easy to fix?
Also we don't ever mention why we use graphs. Is it common practice for point set registration?

Reply:
@Charlottez112 , I'm not sure what you mean by using graphs- which part of the code are you referring to? 

Regarding the easy fix, I do think there are much more elegant and globally minimal ways of solving this matching problem. Because of the thresholding, my method will never give a "false" match (as defined by the threshold criterion) between environments, but there are edge cases where it could fail to find the match, for the reasons Vyas was saying. What I could have done originally, for example, was implement a solution to the [assignment problem](https://en.wikipedia.org/wiki/Assignment_problem) like the [Hungarian algorithm](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linear_sum_assignment.html#scipy.optimize.linear_sum_assignment). One could then check if each matched vector pair that minimizes the cost function, the sum of distances between all matched vector pairs, satisfies the threshold criterion. I have no idea about the timing of this type of solution but that would be more ideal than what I implemented, probably.